make learners picklable

basnijholt opened the pull request:
## Description

I realized that in some cases it's very useful to pickle learners. For example to send it over the network when parallelizing code.

With these changes, most learners become picklable.

## Checklist

- [x] Fixed style issues using `pre-commit run --all` (first install using `pip install pre-commit`)
- [ ] `pytest` passed

## Type of change

*Check relevant option(s).*

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

basnijholt commented:
@akhmerov and @jbweston, I wonder if we should also pickle `adaptive.__version__` and warn the user that if they are loading a learner that is pickled with another version, that there is no guarantee that stuff's correct?

basnijholt commented:
@akhmerov, good idea, I've added that in https://github.com/python-adaptive/adaptive/pull/264/commits/54017f38e01e509c256565abe32b1ddff03eed3d.

akhmerov commented on adaptive/tests/test_pickling.py:
```diff
@@ -0,0 +1,169 @@
+        del learner
+
+    learner_loaded = serializer.loads(learner_bytes)
+    assert learner_loaded.npoints >= 10
```
Why not an exact equality here?

basnijholt replied:
Fixed!

akhmerov commented:
@basnijholt any idea why tests fail?

basnijholt commented:
The failure is intermittent and unrelated to these changes, also happens in master sometimes. 

It seems like the `distributed` test keeps getting stuck on Windows and MacOS.

akhmerov commented:
But it seems like serialization tests are failing?

![image](https://user-images.githubusercontent.com/2069677/79159902-3643a700-7dd9-11ea-8746-f292132613f6.png)

basnijholt commented:
Oh, you are right! I assumed it were the same failures I had seen before.

I think that is fixed by https://github.com/python-adaptive/adaptive/pull/264/commits/bc652133b91673d5f6e14bde15810e5846ee6052.

akhmerov commented on adaptive/tests/test_pickling.py:
```diff
@@ -0,0 +1,173 @@
+
+    learner = learner_type(f, **learner_kwargs)
+    if learner_type is Learner1D:
+        learner._recompute_losses_factor = 1
```
If I understand correctly, here we're saying "we're only computing stuff approximately, so let's change that to make pickling work". I don't think that's the right way to go.

I think pickling is supposed to also store the internal state, so that we shouldn't need to tweak the learner behavior like that.

basnijholt replied:
I don't think it is necessary to require that an unpickled learner is identical to the one before. In this case, the loaded learner has a better approximation of the loss because it is able to load all points at once.

Storing all tiny details will probably mean that with the tiniest of changes (to private API), one is not able to load saved learners from a previous version.

akhmerov replied:
> In this case, the loaded learner has a better approximation of the loss because it is able to load all points at once.

That is a possible reason to deviate from the standard behavior, however, I'd also like to hear @jbweston's opinion—I still think that this is dangerously unconventional.

> Storing all tiny details will probably mean that with the tiniest of changes (even to private API), one is not able to load saved learners from a previous version.

I don't think cross-version pickling is guaranteed by anything.

basnijholt replied:
Even today I loaded data saved with Adaptive more than three years ago and it loaded without any complaint, which made me very happy. 

One other thing to keep in mind is that, even though this might not be the final answer, it's still better than the current broken state! Right now picking works for some Learners while others have a broken state after loading.

akhmerov replied:
That's the difference between data formats and pickling (preserving live objects).

One should be stable, the other should be exact.

jbweston replied:
It is generally true that people expect `pickle.loads(pickle.dumps(x))` to be as close to the identity as possible. If we start deviating from this it is IMO almost certain that we (or one of our users) will get surprised by this at some point in the future.

I would prefer that if what someone wants to do is "create a learner with the same data" then they should export the data from the learner, pickle that, and then load it back in later.

jbweston replied:
I am going to counter the "but this is functionally the same thing" response with "no". 

There are certain semantics associated with pickling/unpickling that people expect.

We do not set up the expectation that exporting/importing data from a learner has any given semantics beyond the most trivial (e.g. that after the import the learner has all the data specified).

basnijholt replied:
OK fair enough, I've made the appropriate changes such that the data structures are the same!

I've run the test 10_000 times and it always passes now! 🎉

basnijholt commented on adaptive/tests/test_pickling.py:
```diff
@@ -0,0 +1,169 @@
+        f = f_for_pickle_datasaver  # noqa: F811
+
+    _learner = Learner1D(f, bounds=(-1, 1))
+    learner = DataSaver(_learner, arg_picker=operator.itemgetter("y"))
```
Completely refactored this.

basnijholt commented on adaptive/tests/test_pickling.py:
```diff
@@ -0,0 +1,169 @@
+    "serializer", serializers,
+)
+def test_serialization_for_datasaver(serializer):
+    def f(x):
```
Everything is now specified in the `learners` list.

basnijholt commented on adaptive/tests/test_pickling.py:
```diff
@@ -0,0 +1,169 @@
+    data = learner.data
+
+    if serializer is not pickle:
+        # With pickle the functions are only pickled by reference
```
After refactor this special treatment of pickle isn't needed anymore.

basnijholt commented on adaptive/tests/test_pickling.py:
```diff
@@ -0,0 +1,169 @@
+
+    if serializer is pickle:
+        # f from the local scope cannot be pickled
+        f = f_for_pickle  # noqa: F811
```
Done!

basnijholt commented on adaptive/tests/test_pickling.py:
```diff
@@ -0,0 +1,169 @@
+    """Test serializing a learner using different serializers."""
+
+    def f(x):
+        return random.random()
```
Fixed.

basnijholt commented on adaptive/tests/test_pickling.py:
```diff
@@ -0,0 +1,169 @@
+
+
+def f_for_pickle(x):
+    return 1
```
Fixed.

basnijholt commented on adaptive/tests/test_pickling.py:
```diff
@@ -0,0 +1,169 @@
+@pytest.mark.parametrize(
+    "serializer", serializers,
+)
+def test_serialization_for_balancing_learner(serializer):
```
Fixed.

basnijholt commented on adaptive/learner/learnerND.py:
```diff
@@ -1176,3 +1169,23 @@ def _get_plane_color(simplex):
+
+    def _set_data(self, data):
+        if data:
+            self.tell_many(*zip(*data.items()))
```
You are right!

I will do the work to make the LearnerND pickleable in https://github.com/python-adaptive/adaptive/pull/272.

basnijholt commented on adaptive/learner/learnerND.py:
```diff
@@ -1176,3 +1169,23 @@ def _get_plane_color(simplex):
+    def __setstate__(self, state):
+        function, bounds, loss_per_simplex, data = state
+        self.__init__(function, bounds, loss_per_simplex)
+        self._set_data(data)
```
I will do the work to make the LearnerND pickleable in https://github.com/python-adaptive/adaptive/pull/272.

tlaeven commented:
Using a cached function doesn't work in this branch, is this intentional?

#### Minimal example:
```python
import adaptive
from functools import lru_cache
adaptive.notebook_extension()

@lru_cache()
def g(x):
    return x**2

def f(x):
    return x-g(0)

learner = adaptive.Learner1D(f, [-1,1])

runner = adaptive.Runner(learner)
runner.live_info()
```

Raises an error:
`loky.process_executor.BrokenProcessPool: A task has failed to un-serialize. Please ensure that the arguments of the function are all picklable.`

basnijholt commented:
@tlaeven, this is actually unrelated to these changes. The code you posted never worked, and has nothing to do with Adaptive:
```python
from functools import lru_cache
import loky
from concurrent.futures import ProcessPoolExecutor

@lru_cache()
def g(x):
    return x ** 2

def f(x):
    return x - g(0)

# Before
with ProcessPoolExecutor() as ex:
    fut = ex.submit(f, 0)
    try:
        fut.result()
    except Exception as e:
        print(f"ProcessPoolExecutor failed: {e}")

# Now
ex = loky.get_reusable_executor()
fut = ex.submit(f, 0)
try:
    fut.result()
except Exception as e:
    print(f"loky failed: {e}")
```

basnijholt commented:
@jbweston, I've fixed all that was brought up.

The `LearnerND` takes some more work, so I moved those commits to https://github.com/python-adaptive/adaptive/pull/272.

I think it's ready to merge. Would you take a final look?
